### ISSUE TITLE
Local provider extension

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#5024]
**How to categorize this issue?**

/area dev-productivity testing open-source
/kind enhancement

## What? 💡 

We would like to have a Gardener extension that implements the [Gardener Extension contract](https://github.com/gardener/gardener/blob/master/docs/extensions/overview.md) ([GEP-1](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md)) without requiring any external infrastructure.
The only requirement should be a local kubernetes cluster like KinD, k3d or minikube.

(basically a proper implementation of https://github.com/timebertt/gardener-mock-provider-poc)

## Why? ❓ 

- dev experience: develop gardener completely on your dev machine without any infrastructure credentials/resources required (think costs and productivity) 💰🚀 
- testing: run fast and cost efficient e2e tests – locally and in pipelines (e.g. PR verification) 💎🛡️
- open source: quick and easy setup for first evaluation of gardener and good basis for first contributions 🤝🤹🥳 

## How? ⚙️ 

The basic idea is to spin up pods on the local cluster, that act as shoot worker machines (basically running kubelet + container runtime in a container). Therefore, we reuse the kind node image.

## Status 🚧 

The base implementation was done as part of a hackathon beginning of November, and now we want to finish the implementation to make it ready for usage in development, testing and the open source world.
The first version of the extension will have some limitations (e.g. only single-node shoots, no owner DNSRecords/checks, no LBs for shoots), but it will be enough for getting some first experience with it and enhancing it along the way.

Current work is done on the [local-extension](https://github.com/gardener/gardener/tree/local-extension) branch.
It is still under development and undocumented, but can be tried out already.
This issue serves as a tracking point for open points, issues, etc. for the first version.
You're welcome to provide feedback here when trying it out!

### Open points

- [x] Implement `Infrastructure`, `ControlPlane`, `Worker` resources
- [x] Implement `Network` API and deploy `kindnet`
- [x] Implement `DNSRecord` support (additional coredns deployment for in-cluster DNS, manipulate `/ect/hosts` for making records available on host machine)
- [x] Implement support for `ReversedVPN`
- [x] Test `APIServerSNI=true` (expose the `istio-ingress` deployment via `NodePort` mapping on kind cluster)
- [x] Implement webhook for `CREATE pod` to inject custom coredns into `/etc/resolv.conf` for machine pods
- [x] Upstream fixes in `master` branch:
    - [x] https://github.com/gardener/gardener/issues/4968
    - [x] https://github.com/gardener/gardener/pull/4994
    - [x] https://github.com/gardener/gardener/issues/5001
- Worker node
    - [x] Move image `Dockerfile` into `gardener/gardener` and build via GitHub actions
    - [x] install nerdctl, add a small wrapper disguising as docker, make it work for `docker {pull,cp,info}` in cloud-config-downloader (hack as long as https://github.com/gardener/gardener/issues/4673 is not done)
- Enable e2e-test (shoot creation + deletion) in pipeline and easy setup for devs
    - [x] Create hack script for creating and deleting kind cluster
    - [x] Create hack script for create and delete shoot
    - [x] Kustomization configs for manifests
    - [x] Create new guide for setting things up, excalidraw diagram
      - [x] document limitations
    - [x] mechanism for deploying Gardener components into kind cluster (preferably using the existing helm charts)
      - [x] Cache Go modules in local Dockerfile
      - [x] Remove custom tagging for built images
      - [x] Use TLS for local etcd and Gardener control plane
      - [x] Build `gardener-admission-controller` and activate `SeedAuthorizer`
      - [x] Deploy local extension via `ControllerRegistration` in `skaffold.yaml`
    - [x] move https://github.com/rfranzke/machine-controller-manager-provider-local to gardener org
      - also move node image + GitHub Action into the same repository
      - build mcm-provider-local with GitHub Actions
    - Pipeline integration
        - [x] Prow experiment
        - [x] setup prow for production use: https://github.com/gardener/ci-infra
        - [x] add postsubmit job for e2e test on `local-extension` branch: https://github.com/gardener/ci-infra/pull/10, [successful example run](https://prow.gardener.cloud/view/gs/gardener-prow/logs/post-gardener-e2e-kind/1466692402788962304)
        - [x] add presubmit and periodic job for e2e test: https://github.com/gardener/ci-infra/pull/20
- Miscellaneous
    - [x] Improve `type=worker` label on worker pod (@timebertt suggested `app=machine`)
    - [x] Introduce annotation to allow skipping the infrastructure cleanup sleep step in the gardenlet flow
    - [x] Rebase branch on top of `master` once #4813 got merged
      - [x] Disable owner checks via new `Seed` setting
      - [x] Adapt infrastructure cleanup annotation to renamed key
      - [x] Disable DWD
    - [x] implement machine image version to docker image tag mapping in CloudProfile
    - [x] Implement shoot webhook for `kube-proxy` mutation properly (decoding instead of string replacement)
    - [x] Drop shoot webhook for `vpn-shoot` (this creates the limitation that `ReversedVPN` must be set to `true` which again creates the limitation that `APIServerSNI` must be set to `true`)
    - [x] Cleanup file structure (e.g., move `kubeconfig` out of `example/provider-local/base`)
    - [x] make dependency-watchdog optional via seed setting: https://github.com/gardener/gardener/pull/5075
    - [x] consider adding `Seed.spec.secretRef` to allow test framework to dump stuff in case of test failures
    - [x] prevent creating shoot client in test framework via command line flag
    - [x] check why `make tear-down-local-env` hangs
    - [x] check why `make gardener-down` performs the deletion in the wrong order
    - [x] check `E1201 16:18:59.064839   67179 authentication.go:63] "Unable to authenticate the request" err="[x509: certificate signed by unknown authority, verifying certificate SN=283755144066020074164713627359841489438, SKID=, AKID= failed: x509: certificate signed by unknown authority]"` logs in `gardener-apiserver`
    - [x] use RBAC resources with specific kubeconfigs (instead of admin kubeconfig bypassing everything) when running gardener locally against the kind cluster

Optional steps for future work (out of scope for now):
- storage (local-provisioner, see what "kind" uses)
- Implement a `cloud-controller-manager` and deploy it via the `ControlPlane` API
- `BackupBucket`/`BackupEntry` support (consider local backup feature of etcd-backup-restore)
- properly implement machine types (CloudProfile, MCM)
- switch from `kindnet` to different CNI plugin that supports `NetworkPolicies` (to be able to validate netpols in the seed/shoot)

**Comment by danielfoehrKn** [comment]
> mechanism for deploying Gardener components into kind cluster (preferably using the existing helm charts)

You might consider using the landscaper control plane component once it is ready & has a local `make` script.
It can be used completely local without any landscaper deployment and requires only a minimal configuration file (will do all the cert generation automatically). It also just wraps the control plane helm chart.

**Comment by timebertt** [comment]
We already have a mechanism in place for that (yes, the list is not exhaustive). Also it's not just about deploying, it also about the prerequisites: building the images and loading them into the cluster.
We now use skaffold for this, because it is able to orchestrate the full chain from image build to deployment.

**Comment by briantopping** [comment]
Great stuff, thanks Tim! Love that this works on Mac! A couple items on Apple Silicon:
* Install Rosetta
* Manually copy the `amd64` version of Helm to `src/github.com/gardener/gardener/hack/tools/bin`. The Helm install script from their site tries to copy a `arm64` version of Helm, which does not exist.

**Comment by timebertt** [comment]
All tasks for the first version of provider-local tracked in this issue are completed now.
Let's get some experience with it and enhance it further down the line.
Please open dedicated issues for improvements and bugs :)

/close
